source/validate.cpp

```diff
@@ -261,6 +267,143 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  spv_result_t forward_declare_id(uint32_t id) {
+    forward_ids.insert(id);
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t remove_if_fwd_declared(uint32_t id) {
+    forward_ids.erase(id);
+    return SPV_SUCCESS;
+  }
+
+  size_t forward_declared_id_count() { return forward_ids.size(); }
+
+  bool check_declared_id(uint32_t id) {
+    return declared_ids.find(id) != std::end(declared_ids);
+  }
+
+  const uint32_t* get_binary() const { return binary; }
+
+  const uint32_t* get_instruction_ptr() { return program_counter; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(program_counter - binary)},
+                            diagnostic, error_code);
+  }
+};
+
+spv_result_t SSAPass(validation_state_t& _, bool can_have_forward_declared_ids,
+                     const spv_parsed_instruction_t* inst) {
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  for (int i = 0; i < inst->num_operands; i++) {
+    spv_parsed_operand_t operand = inst->operands[i];
+    spv_operand_type_t type = operand.type;
+    const uint32_t* operand_ptr = inst_ptr + operand.offset;
+
+    auto ret = SPV_ERROR_INTERNAL;
+    switch (type) {
+      case SPV_OPERAND_TYPE_RESULT_ID: {
+        _.remove_if_fwd_declared(*operand_ptr);
+        ret = _.declare_id(*operand_ptr);
+        break;
+      }
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_ID: {
+        if (can_have_forward_declared_ids) {
+          if (_.check_declared_id(*operand_ptr)) {
+            ret = SPV_SUCCESS;
+          } else {
+            ret = _.forward_declare_id(*operand_ptr);
+          }
+        } else {
+          if (_.check_declared_id(*operand_ptr)) {
+            ret = SPV_SUCCESS;
+          } else {
+            ret = _.diag(SPV_ERROR_INVALID_ID) << "ID " << *operand_ptr
+                                               << " has not been declared";
+          }
+        }
+        break;
+      }
+      default:
+        ret = SPV_SUCCESS;
+        break;
+    }
+    if (SPV_SUCCESS != ret) {
+      return ret;
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t pushInstructions(void* user_data,
+                              const spv_parsed_instruction_t* const inst) {
+  validation_state_t& _ = *(reinterpret_cast<validation_state_t*>(user_data));
+  _.set_instruction_words(inst->words);
+
+  const uint32_t* inst_ptr = _.get_instruction_ptr();
+
+  vector<SpvOp_> instructions_with_forward_ids = {
+      SpvOpExecutionMode, SpvOpEntryPoint, SpvOpPhi, SpvOpFunctionCall,
+      SpvOpName, SpvOpMemberName, SpvOpDecorate, SpvOpMemberDecorate,
+      SpvOpGroupDecorate};
+
+  bool can_have_forward_declared_ids = false;
+  for (auto& instruction : instructions_with_forward_ids) {
+    can_have_forward_declared_ids |= inst->opcode == instruction;
+  }
+
+  auto err = SSAPass(_, can_have_forward_declared_ids, inst);
+
+  if (err != SPV_SUCCESS) {
+    return err;
+  }
+
+  return SPV_SUCCESS;
```

Comment:
I believe L398 - L404 can be combined into one instruction `return SSAPass(...)`?